adding profiles.json documentation

cinnamon-msft opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request
Added documentation for profiles.json file.

<!-- Other than the issue solved, is this relevant to any other issues/existing PRs? --> 
## References

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist
* [ ] Closes #xxx
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed
* [ ] Requires documentation to be updated
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments

DHowett-MSFT commented on doc/ProfilesJSONDocumentation.md:
```diff
@@ -0,0 +1,39 @@
+| fontFace | String | Name of the font face used in the profile. |
+| fontSize | Integer | Sets the font size. |
+| foreground | String | Sets the foreground color of the profile. Overrides `foreground` set in color scheme if `colorscheme` is set. |
+| guid | String | Unique identifier of the profile. |
```
Hmmm. Do we need some indicator about what properties are required? perhaps we should even sort them, required first?

the `guid` is utterly required. `colorTable` isn't.
for a color scheme, the name is required.

Maybe this makes sense:

| Prop | Req | Type | Doc |
| - | - | - | - |
| `guid` | *required* | GUID String | etc
| `colorScheme` | | string | color scheme name

where `guid` unequivocally comes first as it's like the "primary key" of the type

mKay00 replied:
How about making required properties bold and insert an empty line before Arrays (profiles, schemes) to emphasize that these are array objects with their properties below.

zadjii-msft replied:
I agree that we need to make the optional properties clearly distinct from the required ones. I'm a little worried that bold might not be enough to differentiate, since the **globals**, **profiles**, etc, headers are all bolded was well.

DHowett-MSFT commented on doc/ProfilesJSONDocumentation.md:
```diff
@@ -0,0 +1,39 @@
+| fontFace | String | Name of the font face used in the profile. |
+| fontSize | Integer | Sets the font size. |
+| foreground | String | Sets the foreground color of the profile. Overrides `foreground` set in color scheme if `colorscheme` is set. |
+| guid | String | Unique identifier of the profile. |
```
for guid specifically we might need to indicate that it's a _guid_ in registry format (`{00000000-0000-0000-0000-000000000000}`)

DHowett-MSFT commented on doc/ProfilesJSONDocumentation.md:
```diff
@@ -0,0 +1,39 @@
+| Property | Type | Description |
+| -------- | ---- | ----------- |
+| **Global** |  | **Properties listed below affect the entire window, regardless of the property settings.** |
+| alwaysShowTabs | Boolean | When set to `true`, tabs are always displayed. When set to `false` and `experimental_showTabsInTitlebar` is set to `false`, tabs only appear after typing <kbd>Ctrl</kbd> + <kbd>T</kbd>. |
```
since these are key names, should we enclose them in literal backticks as well?

zadjii-msft commented on doc/ProfilesJSONDocumentation.md:
```diff
@@ -0,0 +1,39 @@
+| showTerminalTitleInTitlebar | Boolean | When set to `true`, titlebar displays the title of the selected tab. When set to `false`, titlebar displays "Windows Terminal". |
+| **profiles** | **Array[Object]** | **Properties listed below are specific to each unique profile.** |
+| acrylicOpacity | Integer | Sets the transparency of the window for the profile. Accepts values from 0-1. |
+| background | String | Sets the background color of the profile. Overrides `background` set in color scheme if `colorscheme` is set. |
```
Should we call out that the colors should be a hex string, i.e. `#rrggbb`?

zadjii-msft commented on doc/ProfilesJSONDocumentation.md:
```diff
@@ -0,0 +1,39 @@
+| acrylicOpacity | Integer | Sets the transparency of the window for the profile. Accepts values from 0-1. |
+| background | String | Sets the background color of the profile. Overrides `background` set in color scheme if `colorscheme` is set. |
+| closeOnExit | Boolean | When set to `true`, the selected tab closes when `exit` is typed. When set to `false`, the tab will remain open when `exit` is typed. |
+| colorscheme | String | Name of the color scheme used in the profile that is defined under `schemes`. |
```
```suggestion
| colorscheme | String | Name of the terminal color scheme to use. Color schemes are defined under `schemes`. |
```

zadjii-msft commented on doc/ProfilesJSONDocumentation.md:
```diff
@@ -0,0 +1,39 @@
+| background | String | Sets the background color of the profile. Overrides `background` set in color scheme if `colorscheme` is set. |
+| closeOnExit | Boolean | When set to `true`, the selected tab closes when `exit` is typed. When set to `false`, the tab will remain open when `exit` is typed. |
+| colorscheme | String | Name of the color scheme used in the profile that is defined under `schemes`. |
+| colorTable | Array[String] | Array of colors used in the profile if `colorscheme` is not set. Ordering is as follows: `[black, red, green, yellow, blue, magenta, cyan, white, bright black, bright red, bright green, bright yellow, bright blue, bright magenta, bright cyan, bright white]` |
```
Did we change these to be named as well as opposed to the raw table?

zadjii-msft commented on doc/ProfilesJSONDocumentation.md:
```diff
@@ -0,0 +1,39 @@
+| requestedTheme | String | Sets the theme of the tab bar. Possible values: `"light"`, `"dark"`, `"system"` |
+| showTerminalTitleInTitlebar | Boolean | When set to `true`, titlebar displays the title of the selected tab. When set to `false`, titlebar displays "Windows Terminal". |
+| **profiles** | **Array[Object]** | **Properties listed below are specific to each unique profile.** |
+| acrylicOpacity | Integer | Sets the transparency of the window for the profile. Accepts values from 0-1. |
```
We should probably also call out that this is only used when `useAcrylic` is `true`

zadjii-msft commented on doc/ProfilesJSONDocumentation.md:
```diff
@@ -0,0 +1,39 @@
+| snapOnInput | Boolean | When set to `true`, the window will scroll to the command input line when typing. When set to `false`, the window will not scroll when you start typing. |
+| startingDirectory | String | The directory the shell starts in when it is loaded. |
+| useAcrylic | Boolean | When set to `true`, the window will have an acrylic background. When set to `false`, the window will have a plain, untextured background. |
+| **schemes** | **Array[Object]** | **Properties listed below are specific to each color theme.** |
```
```suggestion
| **schemes** | **Array[Object]** | **Properties listed below are specific to each color scheme.** |
```
I want to make sure we keep `schemes` separate from `themes`, because when we do add support for theming, it's going to be a very different thing than just the color schemes :)

zadjii-msft commented on doc/ProfilesJSONDocumentation.md:
```diff
@@ -0,0 +1,39 @@
+| useAcrylic | Boolean | When set to `true`, the window will have an acrylic background. When set to `false`, the window will have a plain, untextured background. |
+| **schemes** | **Array[Object]** | **Properties listed below are specific to each color theme.** |
+| background | String | Sets the background color of the color scheme. |
+| colors | Array[String] | Array of colors used in the color scheme. Ordering is as follows: `[black, red, green, yellow, blue, magenta, cyan, white, bright black, bright red, bright green, bright yellow, bright blue, bright magenta, bright cyan, bright white]` |
```
This was also changed, see #742

cinnamon-msft commented:
I updated the file addressing all comments. For the colorTable, I don't think it was changed from the original array, but I'm not certain.
Also, I don't think we should put default values into the documentation since they'll already be set in the profiles.json file and it would clutter this file.
Lastly, I added documentation for key bindings :)

rbeesley commented:
For something like the booleans, wouldn't it make more sense for them to just have a default instead of making them required? In fact, other than a profile GUID, what couldn't be created on behalf of the user, and even then I don't see why upon loading the settings anything missing should just be populated automatically including generating a GUID automatically... in which case required vs. optional is not a valuable distinction. Furthermore the defaults could be saved in a skeleton template file, used for creating new profiles, with a fallback built into the app for creating the skeleton if it is missing or corrupt.

zadjii-msft commented:
@rbeesley I think what you're thinking of will be done in #754 (when we get to it)

miniksa commented on doc/ProfilesJSONDocumentation.md:
```diff
@@ -0,0 +1,57 @@
+| `showTerminalTitleInTitlebar` | _Required_ | Boolean | When set to `true`, titlebar displays the title of the selected tab. When set to `false`, titlebar displays "Windows Terminal". |
+| `experimental_showTabsInTitlebar` | Optional | Boolean | When set to `true`, the tabs are moved into the titlebar and the titlebar disappears. When set to `false`, the titlebar sits above the tabs. |
+| **profiles** | | **Array[Object]** | **Properties listed below are specific to each unique profile.** |
+| `acrylicOpacity` | _Required_ | Number | When `useAcrylic` is set to `true`, it sets the transparency of the window for the profile. Accepts values from 0-1. |
```
Accepts... floating point values from 0-1?

miniksa commented on doc/ProfilesJSONDocumentation.md:
```diff
@@ -0,0 +1,57 @@
+| `colorscheme` | _Required_ | String | Name of the terminal color scheme to use. Color schemes are defined under `schemes`. |
+| `commandline` | _Required_ | String | Executable used in the profile. |
+| `cursorColor` | _Required_ | String | Sets the cursor color for the profile. Uses hex color format: `"#rrggbb"`. |
+| `cursorShape` | _Required_ | String | Sets the cursor shape for the profile. Possible values: `"vintage"`, `"bar"`, `"underscore"`, `"filledBox"`, `"emptyBox"` |
```
In this case (and the other cases where we say what the string options are), should we provide an example of what each one does either as a description or as a screenshot?

miniksa commented on doc/ProfilesJSONDocumentation.md:
```diff
@@ -0,0 +1,57 @@
+| `guid` | _Required_ | String | Unique identifier of the profile. Written in registry format: `"{00000000-0000-0000-0000-000000000000}"`. |
+| `historySize` | _Required_ | Integer | The number of lines above the ones displayed in the window you can scroll back to. |
+| `name` | _Required_ | String | Name of the profile. Displays in the dropdown menu. |
+| `padding` | _Required_ | String | Sets the padding around the text within the window. Can have three different formats: `"#"` sets the same padding for all sides, `"#, #"` sets the same padding for top-bottom and right-left, and `"#, #, #, #"` sets the padding individually for top, right, bottom, and left. |
```
I am pretty sure that the order is "top, left, bottom, right". I've never seen a rectangle specified with right before left...

miniksa replied:
Same goes for top-bottom and left-right...

cinnamon-msft replied:
I used this for reference: https://www.w3schools.com/cssref/pr_padding.asp

miniksa replied:
OK, but this isn't CSS.

This is the code calling for the thicknesses to be set up (for padding):
https://github.com/microsoft/terminal/blob/master/src/cascadia/TerminalControl/TermControl.cpp#L1338

And here's the function specification:
https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.thicknesshelper.fromlengths#Windows_UI_Xaml_ThicknessHelper_FromLengths_System_Double_System_Double_System_Double_System_Double_

Which actually means we're both wrong. It's Left, Top, Right, Bottom.

miniksa commented on doc/ProfilesJSONDocumentation.md:
```diff
@@ -0,0 +1,57 @@
+| `startingDirectory` | _Required_ | String | The directory the shell starts in when it is loaded. |
+| `useAcrylic` | _Required_ | Boolean | When set to `true`, the window will have an acrylic background. When set to `false`, the window will have a plain, untextured background. |
+| `colorTable` | Optional | Array[String] | Array of colors used in the profile if `colorscheme` is not set. Colors use hex color format: `"#rrggbb"`. Ordering is as follows: `[black, red, green, yellow, blue, magenta, cyan, white, bright black, bright red, bright green, bright yellow, bright blue, bright magenta, bright cyan, bright white]` |
+| `cursorHeight` | Optional | Integer | Sets the height of the cursor. Only works when `cursorShape` is set to `"vintage"`. Accepts values from 25-100. |
```
What does 25-100 mean? Clarify.

(It's the percentage of the line height from the bottom. 25 is 25% up from the bottom of the line, 100% is actually pretty much the same as the "filledBox" option).

miniksa commented on doc/ProfilesJSONDocumentation.md:
```diff
@@ -0,0 +1,57 @@
+| `acrylicOpacity` | _Required_ | Number | When `useAcrylic` is set to `true`, it sets the transparency of the window for the profile. Accepts values from 0-1. |
+| `background` | _Required_ | String | Sets the background color of the profile. Overrides `background` set in color scheme if `colorscheme` is set. Uses hex color format: `"#rrggbb"`. |
+| `closeOnExit` | _Required_ | Boolean | When set to `true`, the selected tab closes when `exit` is typed. When set to `false`, the tab will remain open when `exit` is typed. |
+| `colorscheme` | _Required_ | String | Name of the terminal color scheme to use. Color schemes are defined under `schemes`. |
```
There's an inflight PR to change this to `colorScheme`... #1054 (and below reference)

miniksa commented on doc/ProfilesJSONDocumentation.md:
```diff
@@ -0,0 +1,57 @@
+| `requestedTheme` | _Required_ | String | Sets the theme of the application. Possible values: `"light"`, `"dark"`, `"system"` |
+| `showTerminalTitleInTitlebar` | _Required_ | Boolean | When set to `true`, titlebar displays the title of the selected tab. When set to `false`, titlebar displays "Windows Terminal". |
+| `experimental_showTabsInTitlebar` | Optional | Boolean | When set to `true`, the tabs are moved into the titlebar and the titlebar disappears. When set to `false`, the titlebar sits above the tabs. |
+| **profiles** | | **Array[Object]** | **Properties listed below are specific to each unique profile.** |
```
Global above was capitalized. profiles, schemes, and keybindings are not... intentional?

cinnamon-msft replied:
Intentional - profiles, schemes, and keybindings are lowercase in the profiles.json file whereas Global isn't included in the file.

miniksa replied:
Ah OK.

miniksa commented on doc/ProfilesJSONDocumentation.md:
```diff
@@ -0,0 +1,57 @@
+| `cyan` | _Required_ | String | Sets the color used as ANSI cyan. Uses hex color format: `"#rrggbb"`. |
+| `foreground` | _Required_ | String | Sets the foreground color of the color scheme. Uses hex color format: `"#rrggbb"`. |
+| `green` | _Required_ | String | Sets the color used as ANSI green. Uses hex color format: `"#rrggbb"`. |
+| `name` | _Required_ | String | Name of the color scheme. |
```
I understand that you're alphabetizing the parameters by category (required, then optional)... but it is really awkward to have the `name` property for schemes be buried down here (as opposed to at the top) and to have the `foreground` and `background` properties not next to each other in the list.

cinnamon-msft replied:
I'm happy to rearrange based on property relevance?

zadjii-msft replied:
I'd probably do name, then fg/bg, then the rest of the colors

Summon528 commented on doc/cascadia/SettingsSchema.md:
```diff
@@ -0,0 +1,57 @@
+| `initialRows` | _Required_ | Integer | The number of rows displayed in the window upon first load. |
+| `requestedTheme` | _Required_ | String | Sets the theme of the application. Possible values: `"light"`, `"dark"`, `"system"` |
+| `showTerminalTitleInTitlebar` | _Required_ | Boolean | When set to `true`, titlebar displays the title of the selected tab. When set to `false`, titlebar displays "Windows Terminal". |
+| `experimental_showTabsInTitlebar` | Optional | Boolean | When set to `true`, the tabs are moved into the titlebar and the titlebar disappears. When set to `false`, the titlebar sits above the tabs. |
```
```suggestion
| `showTabsInTitlebar` | Optional | Boolean | When set to `true`, the tabs are moved into the titlebar and the titlebar disappears. When set to `false`, the titlebar sits above the tabs. |
```
I think this is not `experimental_` anymore #918

carlos-zamora commented on doc/cascadia/SettingsSchema.md:
```diff
@@ -0,0 +1,57 @@
+| `requestedTheme` | _Required_ | String | Sets the theme of the application. Possible values: `"light"`, `"dark"`, `"system"` |
+| `showTerminalTitleInTitlebar` | _Required_ | Boolean | When set to `true`, titlebar displays the title of the selected tab. When set to `false`, titlebar displays "Windows Terminal". |
+| `showTabsInTitlebar` | Optional | Boolean | When set to `true`, the tabs are moved into the titlebar and the titlebar disappears. When set to `false`, the titlebar sits above the tabs. |
+| **profiles** | | **Array[Object]** | **Properties listed below are specific to each unique profile.** |
```
Could/Should we separate this into multiple tables? I think it's particularly difficult to see where Global properties end and "profiles" begins (see below). Maybe make "profiles" a global property and reference the "section below", then introduce the format for a "profiles" object.
![image](https://user-images.githubusercontent.com/11050425/59129744-884aaf00-8922-11e9-8dfa-2ab9c0147a7d.png)

carlos-zamora replied:
I'd also do the same for "schemes" and "keybindings" and provide a small blurb introducing them. Like, for "schemes", mention the ColorTool and say once that they are represented in hex color format.

carlos-zamora commented on doc/cascadia/SettingsSchema.md:
```diff
@@ -0,0 +1,57 @@
+| `white` | _Required_ | String | Sets the color used as ANSI white. Uses hex color format: `"#rrggbb"`. |
+| `yellow` | _Required_ | String | Sets the color used as ANSI yellow. Uses hex color format: `"#rrggbb"`. |
+| **keybindings** | | **Array[Object]** | **Properties listed below are specific to each custom key binding.** |
+| `command` | _Required_ | String | The command executed when the associated key bindings are pressed. |
```
We should have a growing list of which "commands" are available/acceptable.

carlos-zamora commented on doc/cascadia/SettingsSchema.md:
```diff
@@ -0,0 +1,57 @@
+| `colorScheme` | _Required_ | String | Name of the terminal color scheme to use. Color schemes are defined under `schemes`. |
+| `commandline` | _Required_ | String | Executable used in the profile. |
+| `cursorColor` | _Required_ | String | Sets the cursor color for the profile. Uses hex color format: `"#rrggbb"`. |
+| `cursorShape` | _Required_ | String | Sets the cursor shape for the profile. Possible values: `"vintage"`, `"bar"`, `"underscore"`, `"filledBox"`, `"emptyBox"` |
```
Could/Should we add an image for each of the cursor shapes?

carlos-zamora commented on doc/cascadia/SettingsSchema.md:
```diff
@@ -0,0 +1,74 @@
+# Profiles.json Documentation
+
+## Global
```
```suggestion
## Globals
```

DHowett-MSFT commented on doc/cascadia/SettingsSchema.md:
```diff
@@ -0,0 +1,74 @@
+
+| Property | Necessity | Type | Description |
+| -------- | ---- | ----------- | ----------- |
+| `alwaysShowTabs` | _Required_ | Boolean | When set to `true`, tabs are always displayed. When set to `false` and `experimental_showTabsInTitlebar` is set to `false`, tabs only appear after typing <kbd>Ctrl</kbd> + <kbd>T</kbd>. |
```
```suggestion
| `alwaysShowTabs` | _Required_ | Boolean | When set to `true`, tabs are always displayed. When set to `false` and `showTabsInTitlebar` is set to `false`, tabs only appear after typing <kbd>Ctrl</kbd> + <kbd>T</kbd>. |
```
